enhancements/microshift/update-and-rollback.md

```diff
@@ -0,0 +1,180 @@
+will restore the previously saved backup, and perform an atomic rollback to the N-1 state.
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [Optional]
+
+#### rpm-ostree Configuration
+
+`rpm-ostree` provides a configuration option to check and stage new updates.
+The file `/etc/rpm-ostreed.conf` provides `AutomaticUpdatePolicy` which defaults to `none`.
+
+This option needs to be set to `stage` which downloads and unpacks the update performing
+any package layering. Only a small amount of work is left to be performed at shutdown time
+via the `ostree-finalize-staged.service` systemd unit.
+
+Furthermore, in order to apply this policy, the `rpm-ostreed-automatic.service` systemd
+service must be enabled.
+
+#### MicroShift's Upgrade Systemd Service
+
+A new systemd service file will be packaged into the MicroShift's main RPM. This systemd
+service will monitor the `/run/ostree/staged-deployment` file that stores the staged update deployment info by using the `"ConditionPathExists"` option. This condition will be the
+trigger to run a custom script that saves the MicroShift state into a data backup file.
+
+#### MicroShift Backup
+
+Leveraging the ability to run any custom script before the new OSTree deployment becomes
+active, allow us to perform the following tasks in order:
+* Stop MicroShift systemd service (which does not cause interruption for applications)
+* Save a tarball file of `/var/lib/microshift` to `/var/lib/microshift-backup`
+* Save snapshots of any persistent information store in volumes
+* Reboot to get the new OSTree deployment enabled
+
+#### MicroShift Rollback
+
+`Greenboot` will perform different checks to verify if the Edge Device is in a healthy state.
+If the result is declared as successful, MicroShift will run normally until a new OSTree
+version is exposed to the device. If the restult is declared as failed, `Greenboot` will
+run custom scripts that will perform the following tasks in order:
+* Stop MicroShift systemd service
+* Restore all data from `/var/lib/microshift-backup` directory
+* Execute `rpm-ostree rollback -r` in order to deploy the previous N-1 state
+
+### Risks and Mitigations
+
+MicroShift does not run the Cluster Version Operator to handle upgrades, and it relies
+on OSTree capabilities for transactional updates and rollbacks. There could be ocasions
+when specific tasks are required to perform a successful upgrade of certain components
+such as deprecated API removal, ETCD major versions bump, etc. In order to mitigate this,
+every MicroShift version will have a specific custom upgrade script to jump to the next
+version.
+
+MicroShift is meant to run on Edge Devices with small footprint. Storing backups could
+lead the device to run out of disk space. To mitigate this, we will only support only
+N-1 states, which translates into saving only the backup of a previous state and not
+several of them.
```

Comment:
I think what you're saying here is that we would only ever have one backup at a time, from the previous state of the host. Is that right?

Reply:
That's what I'm trying to say, yes. In this kind of constained environments, saving dozens of backups from previous states might not be the best option regarding disk space.

Reply:
Yes, we want to limit the number of backups we keep. Let's keep it at 1 for now, as you propose.